7903750: TimeBudget (-tb) does not fulfill its promisses

judovana opened the pull request:
Added warning, when time budget is to small, and final run will run at least two times longer.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903750](https://bugs.openjdk.org/browse/CODETOOLS-7903750): TimeBudget (-tb) does not fulfill its promisses (**Bug** - P4)


### Reviewers
 * [Galder Zamarreño](https://openjdk.org/census#galder) (@galderz - no project role)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jcstress.git pull/161/head:pull/161` \
`$ git checkout pull/161`

Update a local copy of the PR: \
`$ git checkout pull/161` \
`$ git pull https://git.openjdk.org/jcstress.git pull/161/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 161`

View PR using the GUI difftool: \
`$ git pr show -t 161`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jcstress/pull/161.diff">https://git.openjdk.org/jcstress/pull/161.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jcstress/pull/161#issuecomment-2701370671)
</details>

judovana committed:
intitial -tb warning

judovana committed:
print time budfget also for listing

Required two small refactorings to get nice code reused

judovana committed:
added warning if timebudget is to small

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/JCStress.java:
```diff
@@ -85,6 +85,14 @@ public void run() throws Exception {
     }
 
+    private TimeBudget getTimeBudget(ConfigsWithScheduler config) {
+        if (config == null) {
```
By returning null you'll get an NPE when `printOn` is called in next line. If `config == null` you will get an NPE now. Seems like this change does not change anything. So, without that, I don't see a reason to wrap the TimeBudget instantiation in a private method.

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/TimeBudget.java:
```diff
@@ -131,10 +134,46 @@ public void printOn(PrintStream out) {
+        long expectedPerTest = countEta(DEFAULT_PER_TEST_MS);
+        boolean print=false;
+        if (expectedPerTest > budget.milliseconds() * 2l) {
+            System.out.println(" + +++ FATAL - your tests will never finish as expected. They will run much longer ");
```
I don't think you should assume `System.out`. All these prints should be on a given `PrintStream out`, see `printOn()` method.

The formatting of the message does not seem to adhere to other messages. You should see how messages are printed elsewhere in the code and try to use same format? E.g.

```
out.println("FATAL: No matching tests.")
```

Also, about this particular message, if this is fatal jcstress should not run. I don't think this is really happening here. The `print=true` does not cause jcstress not to run.

judovana replied:
Damn, sorry for this.  This was changed globa-wide recently, and I had kept it correctly in all prs. Ibviously I had become less vigilant at the end, sorry. Fixed here and also in one  more PR which was affected.

judovana replied:
> ```
> out.println("FATAL: No matching tests.")
> ```
> 
> Also, about this particular message, if this is fatal jcstress should not run. I don't think this is really happening here. The `print=true` does not cause jcstress not to run.

forget to reply this. Yes. it will die correctly later.

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/TimeBudget.java:
```diff
@@ -131,10 +134,46 @@ public void printOn(PrintStream out) {
+            print=true;
+        }
+        if (expectedPerTest * 2 < budget.milliseconds() * 2l) {
+            System.out.println(" + +++ WARNING - your time budget will not be used. Tests will end much sooner.");
```
See how other warning messages are printed. You should follow similar format e.g.

```
System.out.println("WARNING: Allocation profiling is not available: " + e.getMessage());
```

judovana replied:
true. I had adjusted it so it is `+ WARNING: ...` nad same for `FATAL`:
```
 + FATAL: - your tests will never finish as expected. They will run much longer 
 | For 169716 with concurrency factor of 1 You have requested/been given time budget of: 01:00:00
 | That is ~21 ms per test
 + +++ However the real time will be converging to: 5d+21:25:48 +++
 | You can play with internal properties name(value/eta):
 |   jcstress.timeBudget.defaultPerTestMs(3000ms/5d+21:25:48)
 |   jcstress.timeBudget.minTimeMs(30ms/01:24:51)
 |   jcstress.timeBudget.maxTimeMs(60000ms/117d+20:36:00)
 | Which are setting up the exact times the individual tests are trying to converage to.
 + Use with caution! Test run below 100ms is moreover jeopardize the purpose. And will not squeeze the time as you wish.
```

I had kept the leading `+` so it continue to look aligned and is clea where it starts and ends.  If you wish me to remove also leading `+ ` then a slightly different approach would need to be taken.

galderz replied:
Hmmm I see, that might be ok for multi-line comments, but if I were you I would look for other multi-line outputs and see what is done there. In JMH multi line comments are just printed as they are, e.g. https://github.com/openjdk/jmh/blob/master/jmh-core/src/main/java/org/openjdk/jmh/runner/format/TextReportFormat.java#L237

judovana replied:
Not much similar things to be here.  I Wanted to make clear that that lines are bound together, as that paragraph needs some formatting to be readable.   I had modified it so it is without it, aligned with original time budget :
```
   Time budget:
    Initial completion estimate: 01:60:3600 left; at Mon, 2025-03-24 20:11:16
    Initial test time: 30 ms
    FATAL: your tests will never finish as expected. They will run much longer 
    For 168708 with concurrency factor of 1 You have requested/been given time budget of: 01:60:3600
    That is ~21 ms per test
    However the real time will be converging to: 5d+20:1235:74124 +++
    You can play with internal  properties name(value/eta):
        jcstress.timeBudget.defaultPerTestMs(3000ms/5d+20:1235:74124)
        jcstress.timeBudget.minTimeMs(30ms/01:84:5061)
        jcstress.timeBudget.maxTimeMs(60000ms/117d+03:228:13680)
      Which are setting up the exact times the individual tests are trying to converage to.
      Use with caution! Test run below 100ms is moreover jeopardize the purpose. And will not squeeze the time as you wish.

  Attached the non-interactive output stream.
  Printing the progress line at most every 15000 milliseconds.
```

wdyt? I like the orogonal "table like" slightly more.

galderz replied:
I don't have a strong preference for either, but I think it'd be good to be consistent.

judovana replied:
Fair enough, that is making it 51:49  for this latest revision without +'s :) TY!

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/TimeBudget.java:
```diff
@@ -131,10 +134,46 @@ public void printOn(PrintStream out) {
+            print=true;
+        }
+        if (print) {
+            System.out.println(" | For " + expectedTests + " with concurrency factor of " + getConcurentTestsFactor()
```
If it was me, I would switch these `println` calls for `printf` to make the messages more readable.

judovana replied:
I had rewrote it to printf, but at the end found the println slightly more readable. It seems println is a bit more used then printf in the prject, so unless yuo insists I will keep println.

galderz replied:
No, that's fine to leave as is

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/infra/grading/ReportUtils.java:
```diff
@@ -290,37 +290,36 @@ public static boolean statusToPassed(TestResult result) {
         }
         long hours = TimeUnit.MILLISECONDS.toHours(ms);
         ms -= TimeUnit.HOURS.toMillis(hours);
-
```
Unnecessary

judovana replied:
Gosh. Right yo are! I missed that completely! TY!

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/infra/grading/ReportUtils.java:
```diff
@@ -290,37 +290,36 @@ public static boolean statusToPassed(TestResult result) {
-
         long minutes = TimeUnit.MILLISECONDS.toMinutes(ms);
         ms -= TimeUnit.MINUTES.toMillis(minutes);
-
```
Unnecessary

judovana replied:
Gosh. Right yo are! I missed that completely! TY!

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/infra/grading/ReportUtils.java:
```diff
@@ -290,37 +290,36 @@ public static boolean statusToPassed(TestResult result) {
+        return result;
+    }
 
+    public static String getNiceMsTimeDate(long ms) {
```
I would not use the word `Nice`, what does it mean here? Something like `formatTimeDate` or `showTimeDate` would be fine for me.

judovana replied:
Renamed to formatMsToDaysAndTime, wdyt?

judovana committed:
Replaced System.out by preset out

judovana committed:
renamed getNiceMsTimeDate to formatMsToDaysAndTime

judovana committed:
Used  WARNING: and FATAL: instead of +++ WARNING - and +++ FATAL -

judovana committed:
Removed unnecessary shortening of MS before next conversion

-        ms -= TimeUnit.HOURS.toMillis(hours);
-        ms -= TimeUnit.MINUTES.toMillis(minutes);

judovana committed:
Removed unnecesasary NPE check in getTimeBudget

The NPE workaround was causing NPE to be only delayed for later

judovana commented:
hi @galderz  !  TYVM for eyeball! I had  adjusted to most of your comments, and left few question marks.  Hope that is ok. hank you again!

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/JCStress.java:
```diff
@@ -85,6 +85,11 @@ public void run() throws Exception {
 
+    private TimeBudget getTimeBudget(ConfigsWithScheduler config) {
+        TimeBudget timeBudget = new TimeBudget(config.configs.size(), opts.timeBudget());
+        return timeBudget;
```
```suggestion
        return new TimeBudget(config.configs.size(), opts.timeBudget());
```

judovana replied:
Sure. TY!

galderz reviewed (changes requested):
Just a small comment

judovana committed:
Removed table like formatting for time budget warnning

judovana committed:
directly return new instance without intermediate variable

Co-authored-by: Galder Zamarreño <galder@redhat.com>

galderz commented on jcstress-core/src/main/java/org/openjdk/jcstress/JCStress.java:
```diff
@@ -85,6 +85,11 @@ public void run() throws Exception {
     }
 
+    private TimeBudget getTimeBudget(ConfigsWithScheduler config) {
+        TimeBudget timeBudget = new TimeBudget(config.configs.size(), opts.timeBudget());
```
Remove this line, not necessary now.

judovana replied:
:facepalm:

judovana committed:
Removed duplicated construcion of timebudget

galderz commented:
All good now, thanks @judovana! @shipilev can you give it a final review?

judovana commented:
tyvm!

judovana commented:
loosing faith

judovana commented:
Loosing the faith...

judovana commented:
../live..

judovana commented:
Faith restored!

shipilev commented:
Please pull/rebase to current master :)

judovana commented:
hi! All my PRs rebased.

judovana commented:
faith restored

judovana commented:
Is there hope?

judovana commented:
Loosing the faith...

judovana commented:
/open

judovana commented:
keep_alive

judovana commented:
Need for this still valid

judovana commented:
keep alive

judovana commented:
```
Once upon a midnight dreary, while I pondered, weak and weary,
Over many a quaint and curious volume of forgotten lore—
    While I nodded, nearly napping, suddenly there came a tapping,
As of some one gently rapping, rapping at my chamber door.
“’Tis some visitor,” I muttered, “tapping at my chamber door—
            Only this and nothing more.”
```

judovana commented:
```
      Ah, distinctly I remember it was in the bleak December,
And each separate dying ember wrought its ghost upon the floor.
   Eagerly I wished the morrow;- vainly I had sought to borrow
From my books surcease of sorrow- sorrow for the lost Lenore-
For the rare and radiant maiden whom the angels name Lenore-
                 Nameless here for evermore. 
```

judovana commented:
```
 And the silken, sad, uncertain rustling of each purple curtain
Thrilled me—filled me with fantastic terrors never felt before;
    So that now, to still the beating of my heart, I stood repeating
    “’Tis some visitor entreating entrance at my chamber door—
Some late visitor entreating entrance at my chamber door;—
            This it is and nothing more.”
```

judovana commented:
/open

judovana commented:
```
     Presently my soul grew stronger; hesitating then no longer,
“Sir,” said I, “or Madam, truly your forgiveness I implore;
    But the fact is I was napping, and so gently you came rapping,
    And so faintly you came tapping, tapping at my chamber door,
That I scarce was sure I heard you”—here I opened wide the door;—
            Darkness there and nothing more.
```

judovana commented:
/keepalive
